### PR TITLE
BUG: added check for optional param scipy.stats.spearmanr

### PR DESCRIPTION
**Pull request by evgenyzhurko**
Fix for issue #6841 
`nan_policy='omit'` and `nan_policy='propagate'` correspond to documentation now(Fix for #6530 )

### CODE REVIEW

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -3294,26 +3294,31 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate'):
     """
     a, axisout = _chk_asarray(a, axis)
 
-    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+    a_contains_nan, nan_policy = _contains_nan(a, nan_policy)
 
-    if contains_nan and nan_policy == 'omit':
+    if a_contains_nan and nan_policy == 'omit':
         a = ma.masked_invalid(a)
-        b = ma.masked_invalid(b)
```
Comment (evgenyzhurko) [inline scipy/stats/stats.py]: I think that additional if statement isn't the best solution. 
If we have first parameter with `np.nan` element and second parameter without `np.nan` element, second parameter will be masked such as first parameter(it isn't necessary).

Comment (ev-br) [inline scipy/stats/stats.py]: So it's a return of revenge of https://github.com/scipy/scipy/issues/6134, i.e. whether a and b represent paired values of not. A quick check with R:

```
> cor.test(c(1, 2, 3, 4), c(8, 7, 6, NA), method='spearman')
	Spearman's rank correlation rho

data:  c(1, 2, 3, 4) and c(8, 7, 6, NA)
S = 8, p-value = 0.3333
alternative hypothesis: true rho is not equal to 0
sample estimates:
rho 
 -1 

> cor.test(c(1, 2, 3), c(8, 7, 6), method='spearman')

	Spearman's rank correlation rho

data:  c(1, 2, 3) and c(8, 7, 6)
S = 8, p-value = 0.3333
alternative hypothesis: true rho is not equal to 0
sample estimates:
rho 
 -1 
```

I think it's a good idea to match R's behavior, and have a test with a number-to-number comparison against R. See e.g. `test_kendalltau` in `stats/tests/test_stats.py`

@vigna would you want to offer an opinion here?

Comment (vigna) [inline scipy/stats/stats.py]: I think matching R's behaviour is always a good idea, but all this business of "missing values" seems to me a bit mathematical flakey in the first place (I read R's manual and the logic of possible options and behaviours is appalling).

**Comment by rgommers** [comment]
@evgenyzhurko when you continue on this, can you please remove the first two commits and squash the pep8 commit into the one before it?

**Comment by metaperture** [comment]
Just wanted to note here that this fix would be made obsolete by https://github.com/scipy/scipy/pull/6931.

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -3294,26 +3294,29 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate'):
...
     br = None
     if b is not None:
         b, axisout = _chk_asarray(b, axis)
 
-        contains_nan, nan_policy = _contains_nan(b, nan_policy)
+        b_contains_nan, nan_policy = _contains_nan(b, nan_policy)
+
+        if a_contains_nan or b_contains_nan:
+            if nan_policy == 'propagate':
+                return SpearmanrResult(np.nan, np.nan)
```
Comment (ev-br) [inline scipy/stats/stats.py]: Hmmm. I think it should preserve the output shape for n-dim inputs

```
In [20]: corr, pvals = spearmanr([[1, 2, 3], [5, 6, 7]], [[5, 6, 11], [8, 9, np.nan]], nan_policy='propagate')

In [21]: pvals
Out[21]: nan
```

while with scipy 0.18.1,

```
In [5]:  corr, pvals = spearmanr([[1, 2, 3], [5, 6, 7]], [[5, 6, 11], [8, 9, np.nan]], nan_policy='propagate')

In [6]: pvals
Out[6]: 
array([[ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan]])
```

**Comment by ev-br** [comment]
Looks reasonable to me modulo a shortcut for n-dim inputs and nan_policy='propagate'.  Otherwise looks ready --- would you be able to finish it up @evgenyzhurko ?

@metaperture I think gh-6931 is best considered separately: this one is a well localized fix for a well defined bug, so that we can review and merge this quickly. gh-6931 on the other hand requires a separate effort to review.

File: scipy/stats/stats.py (hunk 2 above)
Comment (evgenyzhurko) [inline scipy/stats/stats.py]: I added some changes that provide n-dim array of np.nan with `propagate` option. In the last change i evaluate correlation and pvaluy no matter of `nan_policy` option and after that i check `nan_policy` option. 

Better solution will be calculation size of output array without calculation of correlation and pvalue but i don't know how to calculate this size (i have idea to use sum a.shape[1] and b.shape[1] but i don't sure in that solution).

**Comment by ev-br** [comment]
LGTM, merged. Thanks @evgenyzhurko, and keep them coming!

**Pull request merged**